stratum/lib/security/credentials_manager.cc

```diff
@@ -1,130 +1,161 @@
+DEFINE_string(server_key_file, "", "Path to gRPC server private key file");
+DEFINE_string(server_cert_file, "", "Path to gRPC server certificate file");
+DEFINE_string(client_key_file, "", "Path to gRPC client key file");
+DEFINE_string(client_cert_file, "", "Path to gRPC client certificate file");
 
 namespace stratum {
 
-CredentialsManager::~CredentialsManager() {}
+using ::grpc::experimental::FileWatcherCertificateProvider;
+using ::grpc::experimental::TlsChannelCredentialsOptions;
+using ::grpc::experimental::TlsServerCredentials;
+using ::grpc::experimental::TlsServerCredentialsOptions;
+
+constexpr unsigned int CredentialsManager::kFileRefreshIntervalSeconds;
+
 CredentialsManager::CredentialsManager() {}
 
+CredentialsManager::~CredentialsManager() {}
+
+::util::StatusOr<std::unique_ptr<CredentialsManager>>
+CredentialsManager::CreateInstance(bool secure_only /*=false*/) {
+  auto instance = absl::WrapUnique(new CredentialsManager());
+  RETURN_IF_ERROR(instance->Initialize(secure_only));
+  return std::move(instance);
+}
+
 std::shared_ptr<::grpc::ServerCredentials>
 CredentialsManager::GenerateExternalFacingServerCredentials() const {
   return server_credentials_;
 }
 
-::util::StatusOr<std::unique_ptr<CredentialsManager>>
-CredentialsManager::CreateInstance() {
-  auto instance_ = absl::WrapUnique(new CredentialsManager());
-  RETURN_IF_ERROR(instance_->Initialize());
-  return std::move(instance_);
+std::shared_ptr<::grpc::ChannelCredentials>
+CredentialsManager::GenerateExternalFacingClientCredentials() const {
+  return client_credentials_;
 }
 
-::util::Status CredentialsManager::Initialize() {
-  if (FLAGS_ca_cert.empty() && FLAGS_server_key.empty() &&
-      FLAGS_server_cert.empty()) {
-    LOG(WARNING) << "Using insecure server credentials";
-    server_credentials_ = ::grpc::InsecureServerCredentials();
+::util::Status CredentialsManager::Initialize(bool secure_only) {
+  // Server credentials.
+  if (FLAGS_ca_cert_file.empty() && FLAGS_server_key_file.empty() &&
+      FLAGS_server_cert_file.empty()) {
+    if (secure_only) {
+      LOG(WARNING) << "No crt/key files provided, cannot initiate gRPC server credentials";
+      server_credentials_ = nullptr;
+    } else {
+      LOG(WARNING) << "No key files provided, using insecure server credentials!";
+      server_credentials_ = ::grpc::InsecureServerCredentials();
+    }
   } else {
-    // Load default credentials
-    ::util::Status status;
-    std::string pem_root_certs_;
-    std::string server_private_key_;
-    std::string server_cert_;
-    status.Update(ReadFileToString(FLAGS_ca_cert, &pem_root_certs_));
-    status.Update(ReadFileToString(FLAGS_server_key, &server_private_key_));
-    status.Update(ReadFileToString(FLAGS_server_cert, &server_cert_));
-    if (!status.ok()) {
-      RETURN_ERROR().without_logging()
-          << "Unable to load credentials: " << status.error_message();
+    // Check if certificate files exist/accessible
```

Comment:
Introduced by me: Lines 70-75 checking if the files exist before accessing them for credential validation